From 638d7cc9c43862ebf7814a09aa7829294093c5ae Mon Sep 17 00:00:00 2001
From: Andrea Arcangeli <aarcange@redhat.com>
Date: Tue, 9 Sep 2014 01:16:31 -0400
Subject: [mm] thp: __get_page_tail_foll() can use get_huge_page_tail()

Message-id: <1410225399-8094-15-git-send-email-aarcange@redhat.com>
Patchwork-id: 93367
O-Subject: [RHEL7.1 PATCH 14/22] mm: thp: __get_page_tail_foll() can use get_huge_page_tail()
Bugzilla: 1135506
RH-Acked-by: Rik van Riel <riel@redhat.com>
RH-Acked-by: Larry Woodman <lwoodman@redhat.com>
RH-Acked-by: Rafael Aquini <aquini@redhat.com>

From: Oleg Nesterov <oleg@redhat.com>

Cleanup. Change __get_page_tail_foll() to use get_huge_page_tail()
to avoid the code duplication.

Signed-off-by: Oleg Nesterov <oleg@redhat.com>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Dave Jones <davej@redhat.com>
Cc: Darren Hart <dvhart@linux.intel.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: Mel Gorman <mgorman@suse.de>
Acked-by: Andrea Arcangeli <aarcange@redhat.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>
Signed-off-by: Andrea Arcangeli <aarcange@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/mm/internal.h b/mm/internal.h
index 1247150..3f5e1ae 100644
--- a/mm/internal.h
+++ b/mm/internal.h
@@ -52,12 +52,9 @@ static inline void __get_page_tail_foll(struct page *page,
   * page_cache_get_speculative()) on tail pages.
   */
  VM_BUG_ON(atomic_read(&page->first_page->_count) <= 0);
- VM_BUG_ON(atomic_read(&page->_count) != 0);
- VM_BUG_ON(page_mapcount(page) < 0);
  if (get_page_head)
   atomic_inc(&page->first_page->_count);
- if (compound_tail_refcounted(page->first_page))
-  atomic_inc(&page->_mapcount);
+ get_huge_page_tail(page);
 }
 
 /*
-- 
1.7.1